crash on compilable code: assertion failed: Cannot emit primitive conversion from Lscala/Function0; to Ljava/lang/String;

scalac compiles but dotc crashes.

$ cat abc.scala
```
class X(val y: String)
class Y(y: => String) extends X(y)
```


$ dotc abc.scala
```
java.lang.AssertionError: assertion failed: Cannot emit primitive conversion from Lscala/Function0; to Ljava/lang/String;
	at scala.Predef$.assert(Predef.scala:165)
	at scala.tools.nsc.backend.jvm.BCodeIdiomatic$JCodeMethodN.emitT2T(BCodeIdiomatic.scala:255)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.adapt(BCodeBodyBuilder.scala:901)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:436)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder$$anonfun$genLoadArguments$1.apply(BCodeBodyBuilder.scala:1003)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder$$anonfun$genLoadArguments$1.apply(BCodeBodyBuilder.scala:1003)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadArguments(BCodeBodyBuilder.scala:1005)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genApply(BCodeBodyBuilder.scala:661)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:338)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genStat(BCodeBodyBuilder.scala:90)
<snipped>
```

--- a/compiler/src/dotty/tools/dotc/transform/ParamForwarding.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ParamForwarding.scala
@@ -55,7 +55,9 @@ class ParamForwarding(thisTransformer: DenotTransformer) {
         stat match {
           case stat: ValDef =>
             val sym = stat.symbol.asTerm
-            if (sym is (ParamAccessor, butNot = Mutable)) {
+            if (sym.is(ParamAccessor, butNot = Mutable) && !sym.info.isInstanceOf[ExprType]) {
+              // ElimByName gets confused with methods returning an ExprType,
+              // so avoid param forwarding if parameter is by name. See i1766.scala
               val idx = superArgs.indexWhere(_.symbol == sym)
               if (idx >= 0 && superParamNames(idx) == stat.name) { // supercall to like-named parameter
                 val alias = inheritedAccessor(sym)

--- a/tests/pos/i1776.scala
+++ b/tests/pos/i1776.scala
@@ -1,0 +1,3 @@
+class X(val y: String)
+class Y(y: => String) extends X(y)
+class Z(z: => String) extends X(z)

